### PR TITLE
[BTC Signing] Do not swallow error in case of missing keys with segwit

### LINKED ISSUE #996: [Bitcoin signing] Incomplete TX created when private key is missing
**Describe the bug**

When signing Bitcoin transations, using SegWit scripts, in the erroneous case when private key is missing, a transaction is returned nontheless, without error.
Expected behaviour is to return failure.

**To Reproduce**

Consider test case BitcoinSigning.SignP2SH_P2WSH, with a slight modification (tests/Bitcoin/TWBitcoinSigningTests.cpp):
Comment out the line adding privte keys to the signing input:

```C++
TEST(BitcoinSigning, SignP2SH_P2WPKH) {
    // Setup input
    Proto::SigningInput input;
    input.set_hash_type(TWBitcoinSigHashTypeAll);
    input.set_amount(200'000'000);
    input.set_byte_fee(1);
    input.set_to_address("1Bp9U1ogV3A14FMvKbRJms7ctyso4Z4Tcx");
    input.set_change_address("1FQc5LdgGHMHEN9nwkjmz6tWkxhPpxBvBU");

    auto utxoKey0 = PrivateKey(parse_hex("eb696a065ef48a2192da5b28b694f87544b30fae8327c4510137a922f32c6dcf"));
    auto pubKey0 = utxoKey0.getPublicKey(TWPublicKeyTypeSECP256k1);
    auto utxoPubkeyHash = Hash::ripemd(Hash::sha256(pubKey0.bytes));
    ASSERT_EQ(hex(utxoPubkeyHash), "79091972186c449eb1ded22b78e40d009bdf0089");
    //////
    //////input.add_private_key(utxoKey0.bytes.data(), utxoKey0.bytes.size());
    //////

    auto redeemScript = Script::buildPayToWitnessPublicKeyHash(utxoPubkeyHash);
    auto scriptHash = Hash::ripemd(Hash::sha256(redeemScript.bytes));
    ASSERT_EQ(hex(scriptHash), "4733f37cf4db86fbc2efed2500b4f4e49f312023");
    auto scriptString = std::string(redeemScript.bytes.begin(), redeemScript.bytes.end());
    (*input.mutable_scripts())[hex(scriptHash)] = scriptString;

    auto utxo0 = input.add_utxo();
    auto utxo0Script = Script(parse_hex("a9144733f37cf4db86fbc2efed2500b4f4e49f31202387"));
    utxo0->set_script(utxo0Script.bytes.data(), utxo0Script.bytes.size());
    utxo0->set_amount(1'000'000'000);
    auto hash0 = DATA("db6b1b20aa0fd7b23880be2ecbd4a98130974cf4748fb66092ac4d3ceb1a5477");
    utxo0->mutable_out_point()->set_hash(TWDataBytes(hash0.get()), TWDataSize(hash0.get()));
    utxo0->mutable_out_point()->set_index(1);
    utxo0->mutable_out_point()->set_sequence(UINT32_MAX);

    {
        // test plan (but do not reuse plan result)
        auto plan = TransactionBuilder::plan(input);
        EXPECT_TRUE(verifyPlan(plan, {1'000'000'000}, 200'000'000, 170));
    }

    // Sign
    auto signer = TransactionSigner<Transaction, TransactionBuilder>(std::move(input)); 
    auto result = signer.sign();

    ASSERT_TRUE(result) << result.error();
    auto signedTx = result.payload();

    Data serialized;
    signer.encodeTx(signedTx, serialized);
    EXPECT_EQ(getEncodedTxSize(signedTx), (EncodedTxSize{251, 142, 170}));
}
```

Result: signer.sign() returns success, and a transaction of size 142 bytes.

**Expected behavior**
sign() should return failure.  Correct transaction is 251 bytes long.

**Additional context**
Hint: error is swallowed in TransactionSigner::sign().

Note: this was found while testing new segwit-compliant fee estimation; bug is pre-existent.

### PR DESCRIPTION
**Pull request by catenocrypt**
## Description

BTC Signing: Do not swallow error in case of missing keys with segwit scripts.  Fixes #996.
BTC fee: Estimation works correctly even with missing keys (previously it reverted to simple estimation).

## Testing instructions

Standard unit tests.

## Types of changes

* Bug fix (non-breaking change which fixes an issue)

## Checklist

- [x] Prefix PR title with `[WIP]` if necessary.
- [x] Add tests to cover changes as needed.
- [x] Update documentation as needed.
- [x] If there is a related Issue, mention it in the description (e.g. Fixes #<issue_number> ).

### CODE REVIEW

File: tests/Decred/TWAnySignerTests.cpp [hunk 1]
```diff
@@ -77,8 +77,8 @@ TEST(TWAnySignerDecred, Plan) {
 
     EXPECT_EQ(plan.amount(), 10000000);
     EXPECT_EQ(plan.available_amount(), 39900000);
-    EXPECT_EQ(plan.fee(), 254);
-    EXPECT_EQ(plan.change(), 29899746);
+    EXPECT_EQ(plan.fee(), 225);
```
Comment (hewigovens) [inline tests/Decred/TWAnySignerTests.cpp:80]: how does it affect Decred?

Comment (hewigovens) [reply tests/Decred/TWAnySignerTests.cpp:80]: no segwit for decred?

Comment (catenocrypt) [reply tests/Decred/TWAnySignerTests.cpp:80]: Decred has some issues -- let's handle that in separate PR.
- For planning, Decred reuses Bitcoin code, but for signing it duplicates it
- BTC planning code now uses BTC signing for fee computation
- BTC signing and Decred signing differ in keyHash hash function used
- BTC signing fails with key not found
- now missing key case is handled properly in estimation, fee is more accurate
- Decred signing works, but as fee is different, tx will be slightly different
The issue is that Decred tx estimation uses BTC signing, which is incompatible.  Let's address this in different PR; also check if redundancy between Decred&BTC code can be reduced.

image

Comment (hewigovens) [reply tests/Decred/TWAnySignerTests.cpp:80]: yes, we can change that later but make sure the fee is enough to broadcast, some chain has  min relay fee, don't want to break it

**Comment by catenocrypt** [comment]
> Can you add a swift test case?

Added test, to see that you have already addad :)

**Comment by catenocrypt** [comment]
Android build timeout, re-triggering

File: tests/Decred/TWAnySignerTests.cpp (hunk 1 above)
Comment (catenocrypt) [reply tests/Decred/TWAnySignerTests.cpp:80]: Tested with iOS app, OK
https://github.com/trustwallet/ios/pull/2574#issuecomment-645330107

**Pull request merged**